### PR TITLE
[DD4hep] Update to master 21 10 2020

### PR DESCRIPTION
**Pull request by ianna**
#### PR description:

* Merge all PRs in one: cms-sw/cmssw#31770, cms-sw/cmssw#31810, cms-sw/cmssw#31821
* Allow to select correct sensitive detectors when their SpecPars defined as regular expressions
* Replace tbb with std containers

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by civanch** [comment]
urgent

**Comment by ianna** [comment]
+1

**Comment by ianna** [comment]
Please ignore failing code checks. This is due changes in the external API.

**Comment by ianna** [comment]
@silviodonato, @dpiparo, @qliphy - please, consider this PR for 11:00 am IB. Due to the external API changes the code checks fail because the comparison is done against the externals associated with master. Because of that the e-mail notification to L2s is not send and it may take quite a while before we collect all the signatures.
This PR does not affect any production workflows, but essential for the DD4hep integration and ongoing work of several people. Thanks!

@smuzaffar - FYI

**Comment by ianna** [comment]
@civanch, @perrotta, @kpedro88  - please, review. Note, the changes are technical.

**Comment by perrotta** [comment]
> @civanch, @perrotta, @kpedro88 - please, review. Note, the changes are technical.

To the attention of:
@cms-sw/alca-l2
@cms-sw/simulation-l2 
@cms-sw/upgrade-l2 
(by the way, reconstruction is not requested to sign)

**Comment by civanch** [comment]
+1

**Comment by civanch** [comment]
@kpedro88 , @christopheralanwest , please, take care about this PR - it is our today bottleneck.

**Comment by smuzaffar** [comment]
test parameters:
  code-checks-tools = cms.week1.PR_852def_CMSSW_11_2_X_slc7_amd64_gcc820/46.0-852def

**Comment by smuzaffar** [comment]
code-checks

**Comment by smuzaffar** [comment]
code-checks

**Comment by smuzaffar** [comment]
@ianna , can you please apply the code checks patch?

**Comment by ianna** [comment]
@smuzaffar - is it a new rule? Can the change be postponed until this PR is integrated? Thanks
```diff
diff --git a/DetectorDescription/DDCMS/plugins/test/DDTestSpecParsFilter.cc b/DetectorDescription/DDCMS/plugins/test/DDTestSpecParsFilter.cc
index 1dba92270a7b..3f9605284236 100644
--- a/DetectorDescription/DDCMS/plugins/test/DDTestSpecParsFilter.cc
+++ b/DetectorDescription/DDCMS/plugins/test/DDTestSpecParsFilter.cc
@@ -65,7 +65,7 @@ void DDTestSpecParsFilter::analyze(const Event&, const EventSetup& iEventSetup)
     }
   });
   std::cout << "*** Check names in a path after filtering:\n";
-  for (auto it : myReg) {
+  for (const auto& it : myReg) {
     if (it.second->hasPath("//ME11AlumFrame")) {
       std::cout << it.first << "\n";
     }
```

**Comment by ianna** [comment]
@smuzaffar - code checks in `CMSSW_11_2_X_2020-10-18-2300` IB are clean.

**Comment by smuzaffar** [comment]
@silviodonato , the change here is trivial and once the code checks passed I would suggest to go ahead and merge this and cmsdist PR for 11h IB

**Comment by smuzaffar** [comment]
@silviodonato , all is good , feel free to merge this and cmsdist pr

**Comment by silviodonato** [comment]
merge

**Comment by ianna** [comment]
+1

resigned for the record

**Comment by civanch** [comment]
+1

**Comment by ianna** [comment]
> -1
> 
> Tested at: [6841d20](https://github.com/cms-sw/cmssw/commit/6841d20ddf6fa55017b7232fb956dafef4e5779a)
> 
> CMSSW: CMSSW_11_2_X_2020-10-21-2300
> SCRAM_ARCH: slc7_amd64_gcc820
> You can see the results of the tests here:
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1db636/10194/summary.html
> 
> I found follow errors while testing this PR
> 
> **Failed tests: Build HeaderConsistency ClangBuild**
> 
> * **Build**:
> 
> I found compilation error when building:
> 
> Copying tmp/slc7_amd64_gcc820/src/Alignment/CocoaApplication/bin/cocoa/cocoa to productstore area:
> Entering library rule at Alignment/CocoaApplication
> >> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/Alignment/CocoaApplication/src/CocoaAnalyzer.cc
> In file included from /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/Alignment/CocoaApplication/src/CocoaAnalyzer.cc:6:
> /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/DetectorDescription/DDCMS/interface/DDCompactView.h: In function 'std::vector cms::getAllParameterValuesFromSpecParSections(const DDSpecParRegistry&, const string&, const string&)':
> /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/DetectorDescription/DDCMS/interface/DDCompactView.h:61:28: error: request for member 'second' in 'mySpecParSection', which is of pointer type 'const dd4hep::SpecPar* const' (maybe you meant to use '->' ?)
>        if (mySpecParSection.second->hasPath(nodePath)) {
>                             ^~~~~~
> /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/DetectorDescription/DDCMS/interface/DDCompactView.h:62:33: error: request for member 'second' in 'mySpecParSection', which is of pointer type 'const dd4hep::SpecPar* const' (maybe you meant to use '->' ?)
>          return mySpecParSection.second->value>(parameterName);
>                                  ^~~~~~
> 
> * **Clang**:
> 
> I found compilation error while trying to compile with clang. Command used:
> 
> ```
> USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 64 COMPILER='llvm compile'
> ```
> 
> >> Entering Package Alignment/CocoaApplication
> >> Compiling  /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/Alignment/CocoaApplication/bin/cocoa.cpp
> Entering library rule at Alignment/CocoaApplication
> >> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/Alignment/CocoaApplication/src/CocoaAnalyzer.cc
> In file included from /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/Alignment/CocoaApplication/src/CocoaAnalyzer.cc:6:
> /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/DetectorDescription/DDCMS/interface/DDCompactView.h:61:27: error: member reference type 'const dd4hep::SpecPar *const' is a pointer; did you mean to use '->'?
>       if (mySpecParSection.second->hasPath(nodePath)) {
>           ~~~~~~~~~~~~~~~~^
>                           ->
> /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/DetectorDescription/DDCMS/interface/DDCompactView.h:61:28: error: no member named 'second' in 'dd4hep::SpecPar'
>       if (mySpecParSection.second->hasPath(nodePath)) {

@smuzaffar - this is due to changes in external API. Somehow the new external is not picked up.

File: SimG4Core/Geometry/src/DDG4ProductionCuts.cc [hunk 1]
```diff
@@ -113,9 +113,9 @@ void DDG4ProductionCuts::dd4hepInitialize() {
 
   for (auto const& it : *dd4hepMap_) {
     for (auto const& fit : specs) {
-      for (auto const& pit : fit->paths) {
+      for (auto const& pit : fit.second->paths) {
         if (dd4hep::dd::compareEqual(dd4hep::dd::noNamespace(it.first.name()), dd4hep::dd::realTopName(pit))) {
```
Comment (ghugo83) [inline SimG4Core/Geometry/src/DDG4ProductionCuts.cc:117]: @ianna Why not dd4hep::dd::compareEqualName here as well? :)
There could be regex in ProdCuts as well?

Comment (ianna) [reply SimG4Core/Geometry/src/DDG4ProductionCuts.cc:117]: @ghugo83 - well spotted! Thanks.

**Comment by kpedro88** [comment]
to be clear, I find the expectation that I should review a PR between 1am-4am to be ridiculous.

**Comment by kpedro88** [comment]
+upgrade
minor API changes in MTD geometry packages

File: Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0"?>
 <DDDefinition>
   <open_geometry/>
-  <close_geometry/>
+  <!--close_geometry/-->
```
Comment (cvuosalo) [inline Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml:4]: `Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml` is automatically generated, is it not? This change will be lost after the next change to 2021 geometry. Is that intended?

Comment (kpedro88) [reply Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml:4]: That's a good point - unit tests may fail now.

**Comment by ianna** [comment]
> to be clear, I find the expectation that I should review a PR between 1am-4am to be ridiculous.

apologies, it will not happen again :-)

File: SimG4Core/Geometry/src/DDG4ProductionCuts.cc (hunk 1 above)
Comment (ianna) [reply SimG4Core/Geometry/src/DDG4ProductionCuts.cc:117]: done in https://github.com/cms-sw/cmssw/pull/31915

File: Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml (hunk 2 above)
Comment (ianna) [reply Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml:4]: done in https://github.com/cms-sw/cmssw/pull/31915

Comment (ianna) [reply Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml:4]: BTW, these tests are not linked to the Geometry subsystem. The unit tests did not fail during integration of this PR, but later, when an IB was done...

Comment (kpedro88) [reply Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml:4]: Indeed; right now, we don't have a good way to make the test from `Configuration/Geometry` depend on changes in xml files in `Geometry/CMSCommonData/data/dd4hep`.

**Pull request merged**